qmail-smtpd.c

```diff
@@ -217,14 +221,44 @@ int flagbarf; /* defined if seenmail */
 stralloc mailfrom = {0};
 stralloc rcptto = {0};
 
+#ifdef SMTPUTF8
+stralloc mfparms = {0};
+void mailfrom_parms(const char *arg)
+{
+  int i;
+  int len;
+
+  len = str_len(arg);
+  mfparms.len=0;
+  i = byte_chr(arg,len,'>');
+  if (i > 4 && i < len) {
+    while (len) {
+      arg++; len--;
+      if (*arg == ' ' || *arg == '\0' ) {
+         if (case_starts(mfparms.s,"SMTPUTF8")) smtputf8 = 1;
+         mfparms.len=0;
+      } else
+        if (!stralloc_catb(&mfparms,arg,1)) die_nomem();
+    }
+  }
+}
+#endif
+
 void smtp_helo(arg) char *arg;
 {
   smtp_greet("250 "); out("\r\n");
   seenmail = 0; dohelo(arg);
 }
 void smtp_ehlo(arg) char *arg;
 {
-  smtp_greet("250-"); out("\r\n250-PIPELINING\r\n250 8BITMIME\r\n");
+  smtp_greet("250-"); out("\r\n250-PIPELINING\r\n");
+#ifdef SMTPUTF8
+  if (env_get("UTF8")) {
+    flagutf8 = 1;
+    out("250-SMTPUTF8\r\n");
+  }
+#endif
```

Comment:
Cheap to check whether the admin wants to advertise `SMTPUTF8` -- let's just always do it.

Reply:
Then let us remove flagutf8

Reply:
If I understand what `flagutf8` is doing here, I'd keep it as a file global, making sure it gets set in exactly one place (probably in the same place where control files get read). I'd keep all the `flagutf8` checks, too -- just expressed in terms of the global boolean rather than the mechanics of how the config gets applied. And then I might call it `utf8_available`, or something that more precisely expresses this is whether the server offers the capability, as distinct from client-requested states like `utf8_requested` or `utf8_negotiated`.

If I don't understand what `flagutf8` is doing here, let's fix that first :-)